### DISCUSSION TITLE
How to speed up valueMap, values, properties query?

### DISCUSSION THREAD

**Discussion by vitaliishandra** [JanusGraph/janusgraph#3638] [Q&A] [answered]
We have queries like this: g.V().has('AccountId', 1).out('Direct'). It is executed few milliseconds and returns ~13000 results. However when I try to query the values using one of the follwoing queries it took about 35-45 seconds to complete:
 g.V().has('AccountId', 1).out('Direct').values('uid')
 g.V().has('AccountId', 1).out('Direct').valueMap()
 g.V().has('AccountId', 1).out('Direct').properties('uid')

Any ideas how it can be improved? query.batch is enabled.
Jnausgraph version 0.6, backend storage: Cassandra.

**Comment by porunov** [comment, marked as answer]
> We have queries like this: g.V().has('AccountId', 1).out('Direct'). It is executed few milliseconds and returns ~13000 results. However when I try to query the values using one of the follwoing queries it took about 35-45 seconds to complete:
>  g.V().has('AccountId', 1).out('Direct').values('uid')
>  g.V().has('AccountId', 1).out('Direct').valueMap()
>  g.V().has('AccountId', 1).out('Direct').properties('uid')
> 
> Any ideas how it can be improved? query.batch is enabled.
> Jnausgraph version 0.6, backend storage: Cassandra.

You need to enable multiquery (it will be enabled by default from JanusGraph version 1.0.0).
It will make values step faster.
That said, valueMap is not going to leverage that optimization until we add that optimization step. See opened issue: https://github.com/JanusGraph/janusgraph/issues/2444

The same optimization should be implemented for "has" step which isn't leveraging an index (issue: https://github.com/JanusGraph/janusgraph/issues/3244 )

**Reply by vitaliishandra** [reply]
@porunov we have query.batch=true and query.limit-batch-size=false. The has step also use graph index by AccountId prop. As we understand the multiquery  processing is enabled.
Here are two execution plans. Does it mean we need to investigate Cassandra metrics and scale if needed to improve performance?
![image](https://user-images.githubusercontent.com/56965286/222417019-ab2b7c0b-e53e-47ef-8e0a-fdce46f406f2.png)

![image](https://user-images.githubusercontent.com/56965286/222418779-feed42ee-3dbd-4937-9f63-fc6f9e4e52bf.png)

**Reply by porunov** [reply]
Something isn't right here I think:
- Strange that `values` perform worse than `valueMap` even so `values` uses multiquery optimization and `valueMap` doesn't. Perhaps your could try increasing `storage.parallel-backend-executor-service.core-pool-size`?
- Overall 200+ seconds to get 12888 properties from 4296 vertices - sounds quite poor unless the resources are limited too much. Not sure why it happens in your case either.

For example, in our small setup (JanusGraph over AstraDB) we easily get about 20k vertices per second with 13 different properties including LIST and SET String properties of different size. I.e. it's about 260,000 properties / second whereas your setup shows about 64 properties / second. It's over 4000x times difference, even so our setup is far from being optimal and our usage is relatively small compared other complex use-cases where they have 1M vertices / second (which would result in about 200,000x times difference depending on properties size and amount). 
All in all, there are many factors which could influence performance: JanusGraph configuration, JVM configuration, machine resources utilization, network throughput and latency, JanusGraph setup mode (i.e. embedded without serialization or separate with serialization), storage database configuration, storage resources utilization, indices, queries, etc.
I would just suggest trying to profile your JVM and look at where queries spend most of the time. It could be that CPU is overused, or storage queries are very slow due to high latency between your JanusGraph node and storage node (if so, try to locate it closer to storage nodes. In modern could providers it usually can be configuration via specifying proximity for your pods).

I could also miss something obvious, but maybe other community members have more thoughts here.

**Reply by porunov** [reply]
Notice, we are using `janusGraph.multiQuery` quite a lot right now due to those optimizations (described here: https://github.com/JanusGraph/janusgraph/discussions/3638#discussioncomment-5177439) are not yet implemented. We plan to switch back to `valueMap` later after that optimization is implemented.
That also could be a reason for such a performance difference.

**Reply by vitaliishandra** [reply]
Actually we have set storage.parallel-backend-executor-service.core-pool-size=1024 and storage.parallel-backend-ops=true. The result is the same. Regarding the valueMap and values duration differences - it can differ from run to run for us, and in average is the same.
We will check out resource consumption

**Reply by porunov** [reply]
> Actually we have set storage.parallel-backend-executor-service.core-pool-size=1024 and storage.parallel-backend-ops=true. The result is the same. Regarding the valueMap and values duration differences - it can differ from run to run for us, and in average is the same. We will check out resource consumption

It's strange then. Do you know average network latency between JanusGraph and storage? In our case network latency is under 1ms usually but if you have higher latency then it could be the case.

**Reply by porunov** [reply]
Also, we are using JanusGraph 1.0.0-rc2 version, but we didn't have problems with previous versions, so I'm not sure it's the case. Nevertheless, you could try using 1.0.0-rc2 version just in case

**Reply by vitaliishandra** [reply]
@porunov  thatnk you for avices. We have two separate JanusGrpah services. One communicate with Scylla, another - with Cassandra. The data is identical. And we don't see the issues with Scylla (the described queries took up to 3sec). So we will continue investigate what is wrong with Cassandra configuration

**Reply by vitaliishandra** [reply]
We found out that we used azurefile storage with StandardLRS type for Cassandra. We changed it to Premium and now queries wtih values/valueMap/properties took up to 1 sec.
